README.md
A mod for [Factorio](https://www.factorio.com) that adds rail bridges.

---
### Features:

Rail bridges allow rail crossings without signals.

3 bridge types for straight and diagonal crossings.

Research Logistics 3 to unlock rail bridges.

Bridges cannot build on top of rails, mine the rails first.

---
### Known issues:

A train going over and under a bridge at the same time will collide with itself.

Trains always render on top of rail bridges.

/editor instant deconstruction does not work on rail bridges. ([72215](https://forums.factorio.com/viewtopic.php?t=72215))


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.0.4
Date: 2019-07-1
  Changes:
    - Now depends on base <= 0.17.55
  Bugfixes:
    - Fixed on_configuration_changed error.
---------------------------------------------------------------------------------------------------
Version: 0.0.3
Date: 2019-07-10
  Features:
    - Improved graphics.
  Bugfixes:
    - Now supports cloned entities.
---------------------------------------------------------------------------------------------------
Version: 0.0.2
Date: 2019-06-23
  Features:
    - Diagonal bridges.
  Bugfixes:
    - Fixed recipe appearing too early.
---------------------------------------------------------------------------------------------------
Version: 0.0.1
Date: 2019-06-16
  Features:
    - Rail bridges allow rail crossings without signals.
    - Research Logistics 3 to unlock rail bridges.


control.lua
local MY_BRIDGES = {
  ["rail-bridge"] = true,
  ["rail-bridge-diagonal-left"] = true,
  ["rail-bridge-diagonal-right"] = true,
}
local MY_RAILS = {
  ["rail-bridge-north"] = true,
  ["rail-bridge-east"] = true,
  ["rail-bridge-rail-1"] = true,
  ["rail-bridge-rail-2"] = true,
  ["rail-bridge-rail-3"] = true,
  ["rail-bridge-rail-4"] = true,
}

function on_init()
  -- Unlock recipe
  for _, force in pairs(game.forces) do
    if force.technologies["logistics-3"].researched then
      force.recipes["rail-bridge"].enabled = true
      force.recipes["rail-bridge-diagonal-left"].enabled = true
      force.recipes["rail-bridge-diagonal-right"].enabled = true
    end
  end
end

function on_configuration_changed(event)
  on_init()
  local changes = event.mod_changes["rail-bridge"]
  if not changes then return end

  -- Move bridge sprite to the rendering system
  if changes.old_version and changes.old_version < "0.0.3" then
    for _, surface in pairs(game.surfaces) do
      for _, bridge in pairs(surface.find_entities_filtered{name = {
        "rail-bridge",
        "rail-bridge-diagonal-left",
        "rail-bridge-diagonal-right",
      }}) do
        draw_sprite(bridge)
      end
    end
  end
end

function on_built(event)
  local entity = event.created_entity or event.entity
  if not entity or not entity.valid then return end
  if not MY_BRIDGES[entity.name] then return end

  -- Align to rail grid
  local mod_x = 0
  local mod_y = 0
  if entity.name == "rail-bridge-diagonal-left"
  or entity.name == "rail-bridge-diagonal-right" then
    if entity.direction % 4 == defines.direction.north then
      mod_x = 0
      mod_y = 1
    else
      mod_x = 1
      mod_y = 0
    end
  end
  local x = entity.position.x
  local y = entity.position.y
  if x % 2 == mod_x then x = x - 1 end
  if y % 2 == mod_y then y = y - 1 end
  if entity.position.x ~= x or entity.position.y ~= y then
    entity.teleport{x, y}
  end

  -- Check collisions
  local delete_rails = {}
  for _, obstacle in pairs(entity.surface.find_entities_filtered{
    type = {"straight-rail", "entity-ghost"},
    area = area_under(entity),
  }) do
    if obstacle.type == "straight-rail"
    and obstacle.force == entity.force
    and obstacle.can_be_destroyed() then
      table.insert(delete_rails, obstacle)
    elseif obstacle.type == "entity-ghost"
    and obstacle.ghost_type == "straight-rail" then
      table.insert(delete_rails, obstacle)
    else
      -- Obstacle can't be destroyed, abort the build!
      refund_entity(entity, event, obstacle)
      return
    end
  end

  -- Destroy rails in the area
  for _, rail in pairs(delete_rails) do
    if entity.name == "rail-bridge" and MY_RAILS[rail] and rail.direction % 2 == 1 then
      -- Straight bridge can't remove diagonal rails
    else
      refund_entity(rail, event)
    end
  end

  -- Create bridge sprite
  draw_sprite(entity)

  -- Create crossing rails
  if entity.name == "rail-bridge" then
    create_rail("rail-bridge-north", entity, defines.direction.north)
    create_rail("rail-bridge-east", entity, defines.direction.east)

  elseif entity.name == "rail-bridge-diagonal-left" then
    if entity.direction % 4 == defines.direction.north then
      create_rail("rail-bridge-rail-1", entity, defines.direction.north, {0, -1})
      create_rail("rail-bridge-rail-2", entity, defines.direction.north, {0, 1})
      create_rail("rail-bridge-rail-1", entity, defines.direction.southwest, {0, -1})
      create_rail("rail-bridge-rail-2", entity, defines.direction.northeast, {0, 1})
    else
      create_rail("rail-bridge-rail-1", entity, defines.direction.east, {-1, 0})
      create_rail("rail-bridge-rail-2", entity, defines.direction.east, {1, 0})
      create_rail("rail-bridge-rail-1", entity, defines.direction.southeast, {-1, 0})
      create_rail("rail-bridge-rail-2", entity, defines.direction.northwest, {1, 0})
    end

  elseif entity.name == "rail-bridge-diagonal-right" then
    if entity.direction % 4 == defines.direction.north then
      create_rail("rail-bridge-rail-3", entity, defines.direction.north, {0, -1})
      create_rail("rail-bridge-rail-4", entity, defines.direction.north, {0, 1})
      create_rail("rail-bridge-rail-3", entity, defines.direction.southeast, {0, -1})
      create_rail("rail-bridge-rail-4", entity, defines.direction.northwest, {0, 1})
    else
      create_rail("rail-bridge-rail-4", entity, defines.direction.east, {-1, 0})
      create_rail("rail-bridge-rail-3", entity, defines.direction.east, {1, 0})
      create_rail("rail-bridge-rail-4", entity, defines.direction.northeast, {-1, 0})
      create_rail("rail-bridge-rail-3", entity, defines.direction.southwest, {1, 0})
    end
  end
end

function on_entity_cloned(event)
  local entity = event.destination
  if not entity or not entity.valid then return end
  if MY_BRIDGES[entity.name] then
    -- Continue in on_built
    on_built({entity=entity, name=event.name})
  elseif MY_RAILS[entity.name] then
    -- Don't let other mods clone our custom rails
    entity.destroy()
  end
end

function on_destroyed(event)
  local entity = event.entity
  if not entity or not entity.valid then return end
  if not MY_BRIDGES[entity.name] then return end

  -- Remove crossing rails
  for _, rail in pairs(entity.surface.find_entities_filtered {
    type = "straight-rail",
    area = area_under(entity),
    force = entity.force,
  }) do
    if MY_RAILS[rail.name] then
      if entity.name == "rail-bridge" and rail.direction % 2 == 1 then
        -- Straight bridge can't remove diagonal rails
      else
        rail.destroy()
      end
    end
  end
end

function on_gui_opened(event)
  -- Disable bridge gui
  local entity = event.entity
  if not entity or not entity.valid then return end
  if MY_BRIDGES[event.entity.name] then
    game.players[event.player_index].opened = nil
  end
end

function create_rail(name, bridge, direction, position)
  -- Create one of our custom rails
  if not position then position = {0, 0} end
  local rail = bridge.surface.create_entity{
    name = name,
    direction = direction,
    force = bridge.force,
    position = {bridge.position.x + position[1], bridge.position.y + position[2]},
    create_build_effect_smoke = false,
  }
  if rail then rail.destructible = false end
end

function area_under(entity)
  -- Calculate the absolute position of the entity's collision box
  local p = entity.position
  local box = entity.prototype.collision_box
  local dx = (box.right_bottom.x - box.left_top.x) / 2
  local dy = (box.right_bottom.y - box.left_top.y) / 2
  if entity.direction and entity.direction % 4 ~= defines.direction.north then
    dx, dy = dy, dx
  end
  return {{p.x - dx, p.y - dy}, {p.x + dx, p.y + dy}}
end

function refund_entity(entity, build_event, colliding_entity)
  -- Show alert
  local player = nil
  if build_event and build_event.player_index then
    player = game.players[build_event.player_index]
  end
  if player and colliding_entity then
    entity.surface.create_entity{
      name = "flying-text",
      text = {"cant-build-reason.entity-in-the-way", colliding_entity.localised_name},
      position = entity.position,
      render_player_index = build_event.player_index,
    }
  end

  if entity.prototype.items_to_place_this then
    -- Find the item used to place the entity
    local item_count = 0
    local item_name = nil
    if build_event and build_event.stack and build_event.stack.valid_for_read then
      item_name = build_event.stack.name
    end
    for _, item in pairs(entity.prototype.items_to_place_this) do
      if item_name == item.name then
        item_count = item.count
        break
      end
    end
    if item_count == 0 then
      local item = entity.prototype.items_to_place_this[1]
      if item then
        item_name = item.name
        item_count = item.count
      end
    end
    local health = entity.health / entity.prototype.max_health

    -- Return item to player inventory
    if item_count > 0 and player then
      local result = player.insert{name=item_name, count=item_count, health=health}
      item_count = item_count - result
    end

    -- Return item to ground
    if item_count > 0 then
      entity.surface.spill_item_stack(entity.position, {name=item_name, count=item_count, health=health}, false, entity.force, false)
    end
  end

  -- Destroy entity
  entity.destroy{raise_destroy = true}
end

function draw_sprite(entity)
  -- Get sprite name
  local sprite = nil
  if entity.name == "rail-bridge" then
    sprite = "rail-bridge"
  elseif entity.name == "rail-bridge-diagonal-left" then
    if entity.direction % 4 == defines.direction.north then
      sprite = "rail-bridge-ne"
    else
      sprite = "rail-bridge-sw"
    end
  elseif entity.name == "rail-bridge-diagonal-right" then
    if entity.direction % 4 == defines.direction.north then
      sprite = "rail-bridge-nw"
    else
      sprite = "rail-bridge-se"
    end
  end

  -- Draw sprite
  if sprite then
    rendering.draw_sprite{
      sprite = sprite,
      surface = entity.surface,
      target = entity,
      render_layer = "transport-belt",
    }
  end
end

script.on_init(on_init)
script.on_configuration_changed(on_configuration_changed)
script.on_event(defines.events.on_built_entity, on_built)
script.on_event(defines.events.on_robot_built_entity, on_built)
script.on_event(defines.events.script_raised_built, on_built)
script.on_event(defines.events.script_raised_revive, on_built)
script.on_event(defines.events.on_entity_cloned, on_entity_cloned)
script.on_event(defines.events.on_player_mined_entity, on_destroyed)
script.on_event(defines.events.on_robot_mined_entity, on_destroyed)
script.on_event(defines.events.on_entity_died, on_destroyed)
script.on_event(defines.events.script_raised_destroy, on_destroyed)
script.on_event(defines.events.on_gui_opened, on_gui_opened)


prototypes/item.lua
data:extend {
  {
    -- Rail bridge
    type = "item",
    name = "rail-bridge",
    localised_name = {"entity-name.rail-bridge"},
    place_result = "rail-bridge",
    icon = "__rail-bridge__/graphics/icon.png",
    icon_size = 32,
    subgroup = "transport",
    order = "a[train-system]-a[rail]-b[bridge]",
    flags = {},
    stack_size = 10,
  },
  {
    -- Rail bridge diagonal left
    type = "item",
    name = "rail-bridge-diagonal-left",
    localised_name = {"entity-name.rail-bridge-diagonal-left"},
    place_result = "rail-bridge-diagonal-left",
    icon = "__rail-bridge__/graphics/icon-left.png",
    icon_size = 32,
    subgroup = "transport",
    order = "a[train-system]-a[rail]-b[bridge]",
    flags = {},
    stack_size = 10,
  },
  {
    -- Rail bridge diagonal right
    type = "item",
    name = "rail-bridge-diagonal-right",
    localised_name = {"entity-name.rail-bridge-diagonal-right"},
    place_result = "rail-bridge-diagonal-right",
    icon = "__rail-bridge__/graphics/icon-right.png",
    icon_size = 32,
    subgroup = "transport",
    order = "a[train-system]-a[rail]-b[bridge]",
    flags = {},
    stack_size = 10,
  },
}


prototypes/sprite.lua
data:extend{
  {
    -- Underpass for straight bridge
    type = "sprite",
    name = "rail-bridge",
    filename = "__rail-bridge__/graphics/bridge.png",
    priority = "high",
    width = 72,
    height = 32,
    shift = util.by_pixel(0, 34),
    hr_version = {
      filename = "__rail-bridge__/graphics/hr_bridge.png",
      priority = "high",
      width = 144,
      height = 64,
      shift = util.by_pixel(0, 34),
      scale = 0.5,
    }
  },
  {
    -- Underpass for diagonal left bridge
    type = "sprite",
    name = "rail-bridge-ne",
    filename = "__rail-bridge__/graphics/bridge_ne.png",
    priority = "high",
    width = 72,
    height = 72,
    shift = {0, 1.4375},
    hr_version = {
      filename = "__rail-bridge__/graphics/hr_bridge_ne.png",
      priority = "high",
      width = 144,
      height = 144,
      shift = {0, 1.4375},
      scale = 0.5,
    }
  },
  {
    -- Underpass for diagonal left bridge
    type = "sprite",
    name = "rail-bridge-sw",
    filename = "__rail-bridge__/graphics/bridge_sw.png",
    priority = "high",
    width = 72,
    height = 32,
    shift = {-1, 1.0625},
    hr_version = {
      filename = "__rail-bridge__/graphics/hr_bridge_sw.png",
      priority = "high",
      width = 144,
      height = 64,
      shift = {-1, 1.0625},
      scale = 0.5,
    }
  },
  {
    -- Underpass for diagonal right bridge
    type = "sprite",
    name = "rail-bridge-nw",
    filename = "__rail-bridge__/graphics/bridge_nw.png",
    priority = "high",
    width = 72,
    height = 72,
    shift = {0, 1.4375},
    hr_version = {
      filename = "__rail-bridge__/graphics/hr_bridge_nw.png",
      priority = "high",
      width = 144,
      height = 144,
      shift = {0, 1.4375},
      scale = 0.5,
    }
  },
  {
    -- Underpass for diagonal right bridge
    type = "sprite",
    name = "rail-bridge-se",
    filename = "__rail-bridge__/graphics/bridge_se.png",
    priority = "high",
    width = 72,
    height = 32,
    shift = {1, 1.0625},
    hr_version = {
      filename = "__rail-bridge__/graphics/hr_bridge_se.png",
      priority = "high",
      width = 144,
      height = 64,
      shift = {1, 1.0625},
      scale = 0.5,
    }
  },
}


prototypes/entities.lua
local function cut_off_bottom(picture, bottom)
  picture.scale = picture.scale or 1
  local top = -picture.height / 2
  if top < bottom then
    picture.height = bottom - top
    local shift = util.by_pixel(0, (top + picture.height/2) * picture.scale)
    if not picture.shift then
      picture.shift = shift
    else
      picture.shift[1] = picture.shift[1] + shift[1]
      picture.shift[2] = picture.shift[2] + shift[2]
    end
  else
    picture.filename = "__core__/graphics/empty.png"
    picture.width = 1
    picture.height = 1
  end
end

local function copy_and_shift(source, shift)
  local picture = table.deepcopy(source)
  if not picture.shift then picture.shift = {0, 0} end
  if not picture.hr_version.shift then picture.hr_version.shift = {0, 0} end
  picture.shift[1] = picture.shift[1] + shift[1]
  picture.shift[2] = picture.shift[2] + shift[2]
  picture.hr_version.shift[1] = picture.hr_version.shift[1] + shift[1]
  picture.hr_version.shift[2] = picture.hr_version.shift[2] + shift[2]
  return picture
end

local function load_vertical_picture(picture, layer, num)
    picture[layer] = {
      filename = "__rail-bridge__/graphics/vertical_"..layer..".png",
      x = (num - 1) * 128,
      width = 128,
      height = 64,
      variation_count = 1,
      priority = "extra-high",
      flags = {"low-object"},
      hr_version = {
        filename = "__rail-bridge__/graphics/hr_vertical_"..layer..".png",
        x = (num - 1) * 256,
        width = 256,
        height = 128,
        variation_count = 1,
        priority = "extra-high",
        flags = {"low-object"},
        scale = 0.5
      }
    }
end

local function trim_4_to_2_tile_width(picture)
  picture.width = 64
  picture.x = picture.x + 32
  picture.hr_version.width = 128
  picture.hr_version.x = picture.hr_version.x + 64
  return picture
end

-- Horizontal rail for straight bridge
local rail_east = table.deepcopy(data.raw["straight-rail"]["straight-rail"])
rail_east.name = "rail-bridge-east"
rail_east.collision_box = {{-0.4, -0.99}, {0.4, 0.1}}
rail_east.collision_mask = {}
rail_east.minable = nil
rail_east.flags = {"building-direction-8-way", "not-deconstructable", "not-upgradable"}
rail_east.selectable_in_game = false
data:extend{rail_east}

-- Vertical rail for straight bridge
local rail_north = table.deepcopy(rail_east)
rail_north.name = "rail-bridge-north"
rail_north.collision_box = {{-0.4, -0.1}, {0.4, 0.99}}
for _, layer in pairs{"stone_path_background", "stone_path", "ties", "backplates", "metals"} do
  cut_off_bottom(rail_north.pictures.straight_rail_vertical[layer], -14)
  cut_off_bottom(rail_north.pictures.straight_rail_vertical[layer].hr_version, -28)
end
data:extend{rail_north}

-- Rail for diagonal left bridge
local rail_1 = table.deepcopy(rail_east);
rail_1.name = "rail-bridge-rail-1"
rail_1.localised_name = {"entity-name.rail-bridge-diagonal-left"}
rail_1.collision_box = {{-0.1, -0.99}, {0.1, 0.1}}
for _, layer in pairs{"stone_path_background", "stone_path", "ties", "backplates", "metals"} do
  cut_off_bottom(rail_1.pictures.straight_rail_diagonal_right_bottom[layer], 2)
  cut_off_bottom(rail_1.pictures.straight_rail_diagonal_right_bottom[layer].hr_version, 4)
  load_vertical_picture(rail_1.pictures.straight_rail_vertical, layer, 1)
end
data:extend{rail_1}

-- Rail for diagonal left bridge
local rail_2 = table.deepcopy(rail_east);
rail_2.name = "rail-bridge-rail-2"
rail_2.localised_name = {"entity-name.rail-bridge-diagonal-left"}
rail_2.collision_box = {{-0.1, -0.1}, {0.1, 0.99}}
for _, layer in pairs{"stone_path_background", "stone_path", "ties", "backplates", "metals"} do
  cut_off_bottom(rail_2.pictures.straight_rail_diagonal_left_top[layer], -30)
  cut_off_bottom(rail_2.pictures.straight_rail_diagonal_left_top[layer].hr_version, -60)
  load_vertical_picture(rail_2.pictures.straight_rail_vertical, layer, 2)
end
data:extend{rail_2}

-- Rail for diagonal right bridge
local rail_3 = table.deepcopy(rail_east)
rail_3.name = "rail-bridge-rail-3"
rail_3.localised_name = {"entity-name.rail-bridge-diagonal-right"}
rail_3.collision_box = {{-0.1, -0.99}, {0.1, 0.1}}
for _, layer in pairs{"stone_path_background", "stone_path", "ties", "backplates", "metals"} do
  cut_off_bottom(rail_3.pictures.straight_rail_diagonal_left_bottom[layer], 2)
  cut_off_bottom(rail_3.pictures.straight_rail_diagonal_left_bottom[layer].hr_version, 4)
  load_vertical_picture(rail_3.pictures.straight_rail_vertical, layer, 3)
end
data:extend{rail_3}

-- Rail for diagonal right bridge
local rail_4 = table.deepcopy(rail_east)
rail_4.name = "rail-bridge-rail-4"
rail_4.localised_name = {"entity-name.rail-bridge-diagonal-right"}
rail_4.collision_box = {{-0.1, -0.1}, {0.1, 0.99}}
for _, layer in pairs{"stone_path_background", "stone_path", "ties", "backplates", "metals"} do
  cut_off_bottom(rail_4.pictures.straight_rail_diagonal_right_top[layer], -30)
  cut_off_bottom(rail_4.pictures.straight_rail_diagonal_right_top[layer].hr_version, -60)
  load_vertical_picture(rail_4.pictures.straight_rail_vertical, layer, 4)
end
data:extend{rail_4}

-- Straight bridge entity
local bridge = {
  type = "simple-entity-with-force",
  name = "rail-bridge",
  icon = "__rail-bridge__/graphics/icon.png",
  icon_size = 32,
  flags = {"player-creation"},
  collision_mask = {"item-layer", "object-layer", "water-tile"},
  collision_box = {{-0.9, -0.9}, {0.9, 0.9}},
  selection_box = {{-1, -1}, {1, 1}},
  minable = {mining_time = 0.5, result = "rail-bridge"},
  max_health = 300,
  resistances = table.deepcopy(rail_east.resistances),
  corpse = "straight-rail-remnants",
  map_color = table.deepcopy(data.raw["utility-constants"]["default"].chart.rail_color),
  render_layer = "building-smoke",  -- 1 layer below rails
  picture = {layers = {}}
}
-- Add crossing rail graphics
for _, layer in pairs{"stone_path_background", "stone_path", "ties", "backplates", "metals"} do
  table.insert(bridge.picture.layers, table.deepcopy(rail_north.pictures.straight_rail_vertical[layer]))
  table.insert(bridge.picture.layers, table.deepcopy(rail_east.pictures.straight_rail_horizontal[layer]))
end
-- Add bridge graphics
table.insert(bridge.picture.layers, table.deepcopy(data.raw.sprite["rail-bridge"]))
data:extend{bridge}

-- Fake crafting category so we can use assembling-machine as a base entity
data:extend{
  {
    type = "recipe-category",
    name = "rail-bridge-category",
  }
}

-- Diagonal left bridge entity
local diag_left = {
  type = "assembling-machine",
  name = "rail-bridge-diagonal-left",
  icon = "__rail-bridge__/graphics/icon-left.png",
  icon_size = 32,
  subgroup = "transport",
  order = "a[train-system]-a[rail]-b[bridge]",
  flags = {"placeable-neutral", "player-creation", "hide-alt-info"},
  collision_mask = {"item-layer", "object-layer", "water-tile"},
  collision_box = {{-0.55, -1.55}, {0.55, 1.55}},
  selection_box = {{-1, -2}, {1, 2}},
  minable = {mining_time = 0.5, result = "rail-bridge-diagonal-left"},
  max_health = 300,
  resistances = table.deepcopy(rail_east.resistances),
  corpse = "straight-rail-remnants",
  map_color = table.deepcopy(data.raw["utility-constants"]["default"].chart.rail_color),
  crafting_categories = {"rail-bridge-category"},
  crafting_speed = 1,
  energy_usage = "0.0000001W",
  energy_source = {type = "void"},
  off_when_no_fluid_recipe = false,
  fluid_boxes = {{
    base_area = 10,
    base_level = -1,
    production_type = "output",
    pipe_connections = {{type = "output", position = {0, -2}}},
    pipe_picture = {},
    render_layer = "building-smoke",  -- 1 layer below rails
  }},
}
local pipe_picture = diag_left.fluid_boxes[1].pipe_picture
for _, direction in pairs{"north", "east", "south", "west"} do
  pipe_picture[direction] = {layers = {}}
end
-- Add crossing rail graphics
for _, layer in pairs{"stone_path_background", "stone_path", "ties", "backplates", "metals"} do
  for _, direction in pairs{"north", "south"} do
    local offset = -2
    if direction == "north" then offset = 2 end
    table.insert(pipe_picture[direction].layers, trim_4_to_2_tile_width(copy_and_shift(rail_1.pictures.straight_rail_vertical[layer], {0, -1 + offset})))
    table.insert(pipe_picture[direction].layers, trim_4_to_2_tile_width(copy_and_shift(rail_2.pictures.straight_rail_vertical[layer], {0, 1 + offset})))
    table.insert(pipe_picture[direction].layers, copy_and_shift(rail_1.pictures.straight_rail_diagonal_left_bottom[layer], {0, -1 + offset}))
    table.insert(pipe_picture[direction].layers, copy_and_shift(rail_2.pictures.straight_rail_diagonal_right_top[layer], {0, 1 + offset}))
  end
  for _, direction in pairs{"east", "west"} do
    local offset = 2
    if direction == "east" then offset = -2 end
    table.insert(pipe_picture[direction].layers, copy_and_shift(rail_1.pictures.straight_rail_diagonal_right_bottom[layer], {-1 + offset, 0}))
    table.insert(pipe_picture[direction].layers, copy_and_shift(rail_2.pictures.straight_rail_diagonal_left_top[layer], {1 + offset, 0}))
    table.insert(pipe_picture[direction].layers, copy_and_shift(rail_1.pictures.straight_rail_horizontal[layer], {-1 + offset, 0}))
    table.insert(pipe_picture[direction].layers, copy_and_shift(rail_2.pictures.straight_rail_horizontal[layer], {1 + offset, 0}))
  end
end
-- Add bridge graphics
table.insert(pipe_picture["north"].layers, copy_and_shift(data.raw.sprite["rail-bridge-ne"], {0, 2}))
table.insert(pipe_picture["south"].layers, copy_and_shift(data.raw.sprite["rail-bridge-ne"], {0, -2}))
table.insert(pipe_picture["east"].layers, copy_and_shift(data.raw.sprite["rail-bridge-sw"], {-2, 0}))
table.insert(pipe_picture["west"].layers, copy_and_shift(data.raw.sprite["rail-bridge-sw"], {2, 0}))
data:extend{diag_left}

-- Diagonal right bridge entity
local diag_right = table.deepcopy(diag_left)
diag_right.name = "rail-bridge-diagonal-right"
diag_right.icon = "__rail-bridge__/graphics/icon-right.png"
diag_right.minable.result = "rail-bridge-diagonal-right"
pipe_picture = diag_right.fluid_boxes[1].pipe_picture
for _, direction in pairs{"north", "east", "south", "west"} do
  pipe_picture[direction] = {layers = {}}
end
-- Add crossing rail graphics
for _, layer in pairs{"stone_path_background", "stone_path", "ties", "backplates", "metals"} do
  for _, direction in pairs{"north", "south"} do
    local offset = -2
    if direction == "north" then offset = 2 end
    table.insert(pipe_picture[direction].layers, trim_4_to_2_tile_width(copy_and_shift(rail_3.pictures.straight_rail_vertical[layer], {0, -1 + offset})))
    table.insert(pipe_picture[direction].layers, trim_4_to_2_tile_width(copy_and_shift(rail_4.pictures.straight_rail_vertical[layer], {0, 1 + offset})))
    table.insert(pipe_picture[direction].layers, copy_and_shift(rail_3.pictures.straight_rail_diagonal_right_bottom[layer], {0, -1 + offset}))
    table.insert(pipe_picture[direction].layers, copy_and_shift(rail_4.pictures.straight_rail_diagonal_left_top[layer], {0, 1 + offset}))
  end
  for _, direction in pairs{"east", "west"} do
    local offset = 2
    if direction == "east" then offset = -2 end
    table.insert(pipe_picture[direction].layers, copy_and_shift(rail_4.pictures.straight_rail_diagonal_right_top[layer], {-1 + offset, 0}))
    table.insert(pipe_picture[direction].layers, copy_and_shift(rail_3.pictures.straight_rail_diagonal_left_bottom[layer], {1 + offset, 0}))
    table.insert(pipe_picture[direction].layers, copy_and_shift(rail_3.pictures.straight_rail_horizontal[layer], {-1 + offset, 0}))
    table.insert(pipe_picture[direction].layers, copy_and_shift(rail_4.pictures.straight_rail_horizontal[layer], {1 + offset, 0}))
  end
end
-- Add bridge graphics
table.insert(pipe_picture["north"].layers, copy_and_shift(data.raw.sprite["rail-bridge-nw"], {0, 2}))
table.insert(pipe_picture["south"].layers, copy_and_shift(data.raw.sprite["rail-bridge-nw"], {0, -2}))
table.insert(pipe_picture["east"].layers, copy_and_shift(data.raw.sprite["rail-bridge-se"], {-2, 0}))
table.insert(pipe_picture["west"].layers, copy_and_shift(data.raw.sprite["rail-bridge-se"], {2, 0}))
data:extend{diag_right};
